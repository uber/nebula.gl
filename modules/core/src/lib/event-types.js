// @flow

import type { Position } from '../geojson-types.js';

export type DeckGLPick = {
  index: number,
  object: any
};

export type ClickEvent = {
  picks: DeckGLPick[],
  screenCoords: Position,
  groundCoords: Position
};

export type DoubleClickEvent = {
  groundCoords: Position
};

export type StartDraggingEvent = {
  picks: DeckGLPick[],
  screenCoords: Position,
  groundCoords: Position,
  dragStartScreenCoords: Position,
  dragStartGroundCoords: Position
};

export type StopDraggingEvent = {
  picks: DeckGLPick[],
  screenCoords: Position,
  groundCoords: Position,
  dragStartScreenCoords: Position,
  dragStartGroundCoords: Position
};

export type PointerMoveEvent = {
  screenCoords: Position,
  groundCoords: Position,
  picks: DeckGLPick[],
  isDragging: boolean,
  dragStartPicks: ?(DeckGLPick[]),
  dragStartScreenCoords: ?Position,
  dragStartGroundCoords: ?Position,
  sourceEvent: any
};
