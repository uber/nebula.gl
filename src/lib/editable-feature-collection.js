// @flow
import type { GeoJsonGeometry } from '../types';

type FeatureCollection = {
  features: Array<Object>
};

export class EditableFeatureCollection {
  featureCollection: FeatureCollection;

  constructor(featureCollection: FeatureCollection) {
    this.featureCollection = featureCollection;
  }

  getObject(): FeatureCollection {
    return this.featureCollection;
  }

  /**
   * Replaces the position deeply nested withing the given feature's geometry.
   * Works with Point, MultiPoint, LineString, MultiLineString, Polygon, and MultiPolygon.
   *
   * @param featureIndex The index of the feature to update
   * @param positionIndexes An array containing the indexes of the position to replace
   * @param updatedPosition The updated position to place in the result (i.e. [lng, lat])
   *
   * @returns A new `EditableFeatureCollection` with the given position replaced. Does not modify this `EditableFeatureCollection`.
   */
  replacePosition(
    featureIndex: number,
    positionIndexes: Array<number>,
    updatedPosition: Array<number>
  ): EditableFeatureCollection {
    const featureToUpdate = this.featureCollection.features[featureIndex];
    const isPolygonal =
      featureToUpdate.geometry.type === 'Polygon' ||
      featureToUpdate.geometry.type === 'MultiPolygon';

    const updatedCoordinates = immutablyReplacePosition(
      featureToUpdate.geometry.coordinates,
      positionIndexes,
      updatedPosition,
      isPolygonal
    );

    const updatedFeature = {
      ...featureToUpdate,
      geometry: {
        ...featureToUpdate.geometry,
        coordinates: updatedCoordinates
      }
    };

    // Immutably replace the feature being edited in the featureCollection
    const updatedFeatureCollection = {
      ...this.featureCollection,
      features: [
        ...this.featureCollection.features.slice(0, featureIndex),
        updatedFeature,
        ...this.featureCollection.features.slice(featureIndex + 1)
      ]
    };

    return new EditableFeatureCollection(updatedFeatureCollection);
  }

  /**
   * Removes a position deeply nested in a GeoJSON geometry coordinates array.
   * Works with MultiPoint, LineString, MultiLineString, Polygon, and MultiPolygon.
   *
   * @param featureIndex The index of the feature to update
   * @param positionIndexes An array containing the indexes of the postion to remove
   *
   * @returns A new `EditableFeatureCollection` with the given coordinate removed. Does not modify this `EditableFeatureCollection`.
   */
  removePosition(featureIndex: number, positionIndexes: Array<number>): EditableFeatureCollection {
    const featureToUpdate = this.featureCollection.features[featureIndex];
    if (featureToUpdate.geometry.type === 'Point') {
      throw Error(`Can't remove a position from a Point or there'd be nothing left`);
    }
    if (
      featureToUpdate.geometry.type === 'MultiPoint' &&
      featureToUpdate.geometry.coordinates.length < 2
    ) {
      throw Error(`Can't remove the last point of a MultiPoint or there'd be nothing left`);
    }
    const isPolygonal =
      featureToUpdate.geometry.type === 'Polygon' ||
      featureToUpdate.geometry.type === 'MultiPolygon';

    const updatedCoordinates = immutablyRemovePosition(
      featureToUpdate.geometry.coordinates,
      positionIndexes,
      isPolygonal
    );

    const updatedGeometry = {
      ...featureToUpdate.geometry,
      coordinates: updatedCoordinates
    };

    // Handle cases where geometry type is "downgraded"
    downgradeGeometryIfNecessary(updatedGeometry);

    const updatedFeature = {
      ...featureToUpdate,
      geometry: updatedGeometry
    };

    // Immutably replace the feature being edited in the featureCollection
    const updatedFeatureCollection = {
      ...this.featureCollection,
      features: [
        ...this.featureCollection.features.slice(0, featureIndex),
        updatedFeature,
        ...this.featureCollection.features.slice(featureIndex + 1)
      ]
    };

    return new EditableFeatureCollection(updatedFeatureCollection);
  }

  /**
   * Adds a position deeply nested in a GeoJSON geometry coordinates array.
   * Works with MultiPoint, LineString, MultiLineString, Polygon, and MultiPolygon.
   *
   * @param featureIndex The index of the feature to update
   * @param positionIndexes An array containing the indexes of the postion that will preceed the new position
   * @param positionToAdd The new position to place in the result (i.e. [lng, lat])
   *
   * @returns A new `EditableFeatureCollection` with the given coordinate removed. Does not modify this `EditableFeatureCollection`.
   */
  addPosition(
    featureIndex: number,
    positionIndexes: Array<number>,
    positionToAdd: Array<number>
  ): EditableFeatureCollection {
    const featureToUpdate = this.featureCollection.features[featureIndex];

    if (featureToUpdate.geometry.type === 'Point') {
      throw new Error('Unable to add a position to a Point feature');
    }

    const isPolygonal =
      featureToUpdate.geometry.type === 'Polygon' ||
      featureToUpdate.geometry.type === 'MultiPolygon';

    const updatedCoordinates = immutablyAddPosition(
      featureToUpdate.geometry.coordinates,
      positionIndexes,
      positionToAdd,
      isPolygonal
    );

    const updatedFeature = {
      ...featureToUpdate,
      geometry: {
        ...featureToUpdate.geometry,
        coordinates: updatedCoordinates
      }
    };

    // Immutably replace the feature being edited in the featureCollection
    const updatedFeatureCollection = {
      ...this.featureCollection,
      features: [
        ...this.featureCollection.features.slice(0, featureIndex),
        updatedFeature,
        ...this.featureCollection.features.slice(featureIndex + 1)
      ]
    };

    return new EditableFeatureCollection(updatedFeatureCollection);
  }

  replaceGeometry(featureIndex: number, geometry: GeoJsonGeometry) {
    const updatedFeature = {
      ...this.featureCollection.features[featureIndex],
      geometry
    };
    const updatedFeatureCollection = {
      ...this.featureCollection,
      features: [
        ...this.featureCollection.features.slice(0, featureIndex),
        updatedFeature,
        ...this.featureCollection.features.slice(featureIndex + 1)
      ]
    };
    return new EditableFeatureCollection(updatedFeatureCollection);
  }

  addFeature(feature: Object) {
    const updatedFeatureCollection = {
      ...this.featureCollection,
      features: [...this.featureCollection.features, feature]
    };
    return new EditableFeatureCollection(updatedFeatureCollection);
  }

  /**
   * Returns a flat array of positions for the given feature along with their indexes into the feature's geometry's coordinates.
   *
   * @param featureIndex The index of the feature to get edit handles
   */
  getEditHandles(featureIndex: number) {
    let handles = [];

    const geometry = this.featureCollection.features[featureIndex].geometry;

    switch (geometry.type) {
      case 'Point':
        // positions are not nested
        handles = [
          {
            position: geometry.coordinates,
            positionIndexes: [],
            featureIndex,
            type: 'existing'
          }
        ];
        break;
      case 'MultiPoint':
      case 'LineString':
        // positions are nested 1 level
        const includeIntermediate = geometry.type !== 'MultiPoint';
        handles = handles.concat(
          getEditHandles(geometry.coordinates, [], includeIntermediate, featureIndex)
        );
        break;
      case 'Polygon':
      case 'MultiLineString':
        // positions are nested 2 levels
        for (let a = 0; a < geometry.coordinates.length; a++) {
          handles = handles.concat(
            getEditHandles(geometry.coordinates[a], [a], true, featureIndex)
          );
        }
        break;
      case 'MultiPolygon':
        // positions are nested 3 levels
        for (let a = 0; a < geometry.coordinates.length; a++) {
          for (let b = 0; b < geometry.coordinates[a].length; b++) {
            handles = handles.concat(
              getEditHandles(geometry.coordinates[a][b], [a, b], true, featureIndex)
            );
          }
        }
        break;
      default:
        throw Error(`Unhandled geometry type: ${geometry.type}`);
    }

    return handles;
  }
}

function immutablyReplacePosition(
  coordinates: Array<any>,
  positionIndexes: Array<number>,
  updatedPosition: Array<number>,
  isPolygonal: boolean
): Array<any> {
  if (!positionIndexes) {
    return coordinates;
  }
  if (positionIndexes.length === 0) {
    return updatedPosition;
  }
  if (positionIndexes.length === 1) {
    const updated = [
      ...coordinates.slice(0, positionIndexes[0]),
      updatedPosition,
      ...coordinates.slice(positionIndexes[0] + 1)
    ];

    if (
      isPolygonal &&
      (positionIndexes[0] === 0 || positionIndexes[0] === coordinates.length - 1)
    ) {
      // for polygons, the first point is repeated at the end of the array
      // so, update it on both ends of the array
      updated[0] = updatedPosition;
      updated[coordinates.length - 1] = updatedPosition;
    }
    return updated;
  }

  // recursively update inner array
  return [
    ...coordinates.slice(0, positionIndexes[0]),
    immutablyReplacePosition(
      coordinates[positionIndexes[0]],
      positionIndexes.slice(1, positionIndexes.length),
      updatedPosition,
      isPolygonal
    ),
    ...coordinates.slice(positionIndexes[0] + 1)
  ];
}

function immutablyRemovePosition(
  coordinates: Array<any>,
  positionIndexes: Array<number>,
  isPolygonal: boolean
): Array<any> {
  if (!positionIndexes) {
    return coordinates;
  }
  if (positionIndexes.length === 0) {
    throw Error('Must specify the index of the position to remove');
  }
  if (positionIndexes.length === 1) {
    const updated = [
      ...coordinates.slice(0, positionIndexes[0]),
      ...coordinates.slice(positionIndexes[0] + 1)
    ];

    if (
      isPolygonal &&
      (positionIndexes[0] === 0 || positionIndexes[0] === coordinates.length - 1)
    ) {
      // for polygons, the first point is repeated at the end of the array
      // so, if the first/last coordinate is to be removed, coordinates[1] will be the new first/last coordinate
      if (positionIndexes[0] === 0) {
        // change the last to be the same as the first
        updated[updated.length - 1] = updated[0];
      } else if (positionIndexes[0] === coordinates.length - 1) {
        // change the first to be the same as the last
        updated[0] = updated[updated.length - 1];
      }
    }
    return updated;
  }

  // recursively update inner array
  return [
    ...coordinates.slice(0, positionIndexes[0]),
    immutablyRemovePosition(
      coordinates[positionIndexes[0]],
      positionIndexes.slice(1, positionIndexes.length),
      isPolygonal
    ),
    ...coordinates.slice(positionIndexes[0] + 1)
  ];
}

function immutablyAddPosition(
  coordinates: Array<any>,
  positionIndexes: Array<number>,
  positionToAdd: Array<number>,
  isPolygonal: boolean
): Array<any> {
  if (!positionIndexes) {
    return coordinates;
  }
  if (positionIndexes.length === 0) {
    throw Error('Must specify the index of the position to remove');
  }
  if (positionIndexes.length === 1) {
    if (isPolygonal && (positionIndexes[0] < 1 || positionIndexes[0] > coordinates.length - 1)) {
      // TODO: test this case
      throw Error(
        `Invalid position index for polygon: ${
          positionIndexes[0]
        }. Points must be added to a Polygon between the first and last point.`
      );
    }
    const updated = [
      ...coordinates.slice(0, positionIndexes[0]),
      positionToAdd,
      ...coordinates.slice(positionIndexes[0])
    ];
    return updated;
  }

  // recursively update inner array
  return [
    ...coordinates.slice(0, positionIndexes[0]),
    immutablyAddPosition(
      coordinates[positionIndexes[0]],
      positionIndexes.slice(1, positionIndexes.length),
      positionToAdd,
      isPolygonal
    ),
    ...coordinates.slice(positionIndexes[0] + 1)
  ];
}

function downgradeGeometryIfNecessary(geometry: GeoJsonGeometry) {
  switch (geometry.type) {
    case 'LineString':
      downgradeLineStringIfNecessary(geometry);
      break;
    case 'Polygon':
      downgradePolygonIfNecessary(geometry);
      break;
    case 'MultiLineString':
      downgradeMultiLineStringIfNecessary(geometry);
      break;
    case 'MultiPolygon':
      downgradeMultiPolygonIfNecessary(geometry);
      break;
    default:
      // Not downgradable
      break;
  }
}

function downgradeLineStringIfNecessary(geometry: GeoJsonGeometry) {
  if (geometry.coordinates.length === 1) {
    // Only one position left, so convert to a Point
    geometry.type = 'Point';
    geometry.coordinates = geometry.coordinates[0];
  }
}

function downgradePolygonIfNecessary(geometry: GeoJsonGeometry) {
  const polygon = geometry.coordinates;
  const outerRing = polygon[0];
  // If the outer ring is no longer a polygon, convert the whole thing to a LineString
  if (outerRing.length <= 3) {
    geometry.type = 'LineString';
    geometry.coordinates = outerRing.slice(0, outerRing.length - 1);
    return;
  }

  // If any hole is no longer a polygon, remove the hole entirely
  for (let holeIndex = 1; holeIndex < polygon.length; holeIndex++) {
    if (removeHoleIfNecessary(polygon, holeIndex)) {
      // It was removed, so keep the index the same
      holeIndex--;
    }
  }
}

function downgradeMultiLineStringIfNecessary(geometry: GeoJsonGeometry) {
  if (geometry.coordinates.length === 1 && geometry.coordinates[0].length === 1) {
    // Only one position left, so convert to a Point
    geometry.type = 'Point';
    geometry.coordinates = geometry.coordinates[0][0];
    return;
  }
  for (let lineStringIndex = 0; lineStringIndex < geometry.coordinates.length; lineStringIndex++) {
    const lineString = geometry.coordinates[lineStringIndex];
    if (lineString.length === 1) {
      // Only a single position left on this LineString, so remove it (can't have Point in MultiLineString)
      geometry.coordinates.splice(lineStringIndex, 1);
      // Keep the index the same
      lineStringIndex--;
    }
  }
}

function downgradeMultiPolygonIfNecessary(geometry: GeoJsonGeometry) {
  if (geometry.coordinates.length === 1) {
    const outerRing = geometry.coordinates[0][0];
    if (outerRing.length <= 3) {
      geometry.type = 'LineString';
      geometry.coordinates = outerRing.slice(0, outerRing.length - 1);
      return;
    }
  }
  for (let polygonIndex = 0; polygonIndex < geometry.coordinates.length; polygonIndex++) {
    const polygon = geometry.coordinates[polygonIndex];
    const outerRing = polygon[0];

    // If the outer ring is no longer a polygon, remove the whole polygon
    if (outerRing.length <= 3) {
      geometry.coordinates.splice(polygonIndex, 1);
      // It was removed, so keep the index the same
      polygonIndex--;
    }

    for (let holeIndex = 1; holeIndex < polygon.length; holeIndex++) {
      if (removeHoleIfNecessary(polygon, holeIndex)) {
        // It was removed, so keep the index the same
        holeIndex--;
      }
    }
  }
}

function removeHoleIfNecessary(polygon: Array<any>, holeIndex: number) {
  const hole = polygon[holeIndex];
  if (hole.length <= 3) {
    polygon.splice(holeIndex, 1);
    return true;
  }
  return false;
}

function getIntermediatePosition(
  position1: Array<number>,
  position2: Array<number>
): Array<number> {
  const intermediatePosition = [];
  for (let dimension = 0; dimension < position1.length; dimension++) {
    intermediatePosition.push((position1[dimension] + position2[dimension]) / 2.0);
  }
  return intermediatePosition;
}

function getEditHandles(
  coordinates: Array<Array<number>>,
  positionIndexPrefix: Array<number>,
  includeIntermediate: boolean,
  featureIndex: number
) {
  const editHandles = [];
  for (let i = 0; i < coordinates.length; i++) {
    const position = coordinates[i];
    editHandles.push({
      position,
      positionIndexes: [...positionIndexPrefix, i],
      featureIndex,
      type: 'existing'
    });

    if (includeIntermediate && i < coordinates.length - 1) {
      // add intermediate position after every position except the last one
      const nextPosition = coordinates[i + 1];
      editHandles.push({
        position: getIntermediatePosition(position, nextPosition),
        positionIndexes: [...positionIndexPrefix, i + 1],
        featureIndex,
        type: 'intermediate'
      });
    }
  }
  return editHandles;
}
