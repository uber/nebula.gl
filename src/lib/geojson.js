// @flow
import keymirror from 'keymirror';

import Feature from './feature';

export const GeoJsonGeometryTypes = keymirror({
  Point: null,
  LineString: null,
  Polygon: null,
  MultiPoint: null,
  MultiLineString: null,
  MultiPolygon: null
});

export function expandMultiGeometry(
  data: Feature[],
  singleType: string,
  multiType: string,
  createMulti: Function
): { result: Feature[], rejected: Feature[] } {
  const result = [];
  const rejected = [];

  data.forEach(nf => {
    if (nf.geoJson.geometry.type === singleType) {
      result.push(nf);
    } else if (nf.geoJson.geometry.type === multiType) {
      nf.geoJson.geometry.coordinates.forEach((coord, index) => {
        result.push(new Feature(createMulti(coord), nf.style, nf.original, { index }));
      });
    } else {
      rejected.push(nf);
    }
  });

  return { result, rejected };
}

/**
 * Updates a coordinate deeply nested in a GeoJSON geometry coordinates.
 * Works with MultiPoint, LineString, MultiLineString, Polygon, and MultiPolygon.
 *
 * @param coordinates A GeoJSON geometry coordinates array
 * @param indexes An array containing the indexes of the coordinates to replace
 * @param updatedCoordinate The updated coordinate to place in the result (i.e. [lng, lat])
 * @param isPolygonal `true` if `coordinates` is a Polygon or MultiPolygon
 *
 * @returns A new array with the coordinates at the given index replaced. Does not modify `coordinates`.
 */
export function immutablyReplaceCoordinate(
  coordinates: Array<mixed>,
  indexes: Array<number>,
  updatedCoordinate: Array<number>,
  isPolygonal: boolean = false
): Array<mixed> {
  if (!indexes || indexes.length === 0) {
    return coordinates;
  }
  if (indexes.length === 1) {
    const updated = [
      ...coordinates.slice(0, indexes[0]),
      updatedCoordinate,
      ...coordinates.slice(indexes[0] + 1)
    ];

    if (isPolygonal && (indexes[0] === 0 || indexes[0] === coordinates.length - 1)) {
      // for polygons, the first point is repeated at the end of the array
      // so, update it on both ends of the array
      updated[0] = updatedCoordinate;
      updated[coordinates.length - 1] = updatedCoordinate;
    }
    return updated;
  }

  // recursively update inner array
  return [
    ...coordinates.slice(0, indexes[0]),
    immutablyReplaceCoordinate(
      coordinates[indexes[0]],
      indexes.slice(1, indexes.length),
      updatedCoordinate,
      isPolygonal
    ),
    ...coordinates.slice(indexes[0] + 1)
  ];
}
