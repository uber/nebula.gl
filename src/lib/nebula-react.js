// @flow
import React, { Component } from 'react';
import DeckGL from 'deck.gl';
import PropTypes from 'prop-types';

import type { Viewport } from '../types';
import Nebula from './nebula';

type Props = {
  layers: Object[],
  onSelection?: (selectedFeatures: Object[]) => void,
  selectionType?: number,
  onMapMouseEvent?: Function,
  viewport: Viewport,
  logger?: Object,
  children?: any
};

const styles = {
  canvasContainer: {
    pointerEvents: 'none',
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0
  }
};

export default class NebulaReact extends Component<Props> {
  static childContextTypes = {
    viewport: PropTypes.object,
    wmViewport: PropTypes.object,
    nebula: PropTypes.object
  };

  getChildContext() {
    const { nebula } = this;
    const { props: { viewport }, wmViewport } = nebula;
    return { viewport, wmViewport, nebula };
  }

  componentWillMount() {
    this.nebula = new Nebula();
    this.nebula.init(this.props);
    this.nebula.forceUpdate = () => this.forceUpdate();
  }

  componentWillReceiveProps(props: Props) {
    this.nebula.updateProps(props);
  }

  nebula: Nebula;
  mainContainer: ?Object;

  updateAllDeckObjects() {
    this.nebula.updateAllDeckObjects();
  }

  updateDeckObjectsByIds(ids: string[]) {
    this.nebula.updateDeckObjectsByIds(ids);
  }

  render() {
    const { viewport, children } = this.props;
    const { width, height } = viewport;

    return (
      <div style={styles.canvasContainer} ref={div => (this.mainContainer = div)}>
        <DeckGL
          ref={deckgl => this.nebula.setDeck(deckgl)}
          width={width}
          height={height}
          viewports={[this.nebula.wmViewport]}
          layers={this.nebula.getRenderedLayers()}
        />
        {children}
      </div>
    );
  }
}
