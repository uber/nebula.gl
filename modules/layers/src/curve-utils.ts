import hermite from 'cubic-hermite-spline';
import turfDistance from '@turf/distance';
import { lineString } from '@turf/helpers';

import { Feature, MultiLineString, LineString } from 'geojson';

const INTERPOLATION_INTERVAL = 0.005;
const INTERPOLATION_THRESHOLD = 0.001;

function calculateSingleTangent(p0: [number, number], p1: [number, number], d: number): number[] {
  const x = (p1[0] - p0[0]) / d;
  const y = (p1[1] - p0[1]) / d;
  return [x, y];
}

export function generateCurveFromControlPoints(
  line: Feature<MultiLineString>
): Feature<LineString> {
  // calculate knots
  const knots = [0];
  let prev = null;
  let totalDistance = 0;

  const { coordinates: coords } = line.geometry;

  for (let i = 0; i < coords.length; i++) {
    const cur = coords[i];
    if (prev !== null) {
      totalDistance += turfDistance(prev, cur);
      knots.push(totalDistance);
    }
    prev = cur;
  }

  // calculate tangents
  const tangents = [];

  // first tangent
  // @ts-ignore
  tangents.push(calculateSingleTangent(coords[0], coords[1], knots[1] - knots[0]));

  // second to before last
  for (let i = 1; i < coords.length - 1; i++) {
    // @ts-ignore
    const A = calculateSingleTangent(coords[i], coords[i + 1], knots[i + 1] - knots[i]);
    // @ts-ignore
    const B = calculateSingleTangent(coords[i - 1], coords[i], knots[i] - knots[i - 1]);
    const x = (A[0] + B[0]) / 2.0;
    const y = (A[1] + B[1]) / 2.0;
    tangents.push([x, y]);
  }

  // last tangent
  const last = coords.length - 1;
  tangents.push(
    // @ts-ignore
    calculateSingleTangent(coords[last - 1], coords[last], knots[last] - knots[last - 1])
  );

  // generate curve
  const result = [];
  for (let i = 0; i < coords.length; i++) {
    // add control point
    result.push(coords[i]);

    // add interpolated values
    for (let t = knots[i] + INTERPOLATION_INTERVAL; t < knots[i + 1]; t += INTERPOLATION_INTERVAL) {
      if (knots[i + 1] - t > INTERPOLATION_THRESHOLD) {
        // Only add if not too close to a control point (knot = control point)
        result.push(hermite(t, coords, tangents, knots));
      }
    }
  }

  return lineString(result);
}
