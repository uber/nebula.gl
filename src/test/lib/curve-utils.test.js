import { generateCurveFromControlPoints } from '../../lib/curve-utils';

const POLYLINE = {
  type: 'Feature',
  properties: {},
  geometry: {
    type: 'LineString',
    coordinates: [
      [-122.41988182067871, 37.8014343052295],
      [-122.41904497146605, 37.802790657411244],
      [-122.41724252700804, 37.801603850614384],
      [-122.41612672805786, 37.80314669573162]
    ]
  }
};

it('test generateCurveFromControlPoints', () => {
  const result = generateCurveFromControlPoints(POLYLINE);
  expect(result).toMatchSnapshot();
});
