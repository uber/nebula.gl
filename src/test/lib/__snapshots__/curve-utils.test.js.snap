// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generateCurveFromControlPoints 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        -122.41988182067871,
        37.8014343052295,
      ],
      Array [
        -122.41985715431925,
        37.801475724545,
      ],
      Array [
        -122.4198329976012,
        37.80151902179406,
      ],
      Array [
        -122.41980930049003,
        37.801564012608736,
      ],
      Array [
        -122.41978601295123,
        37.80161051262117,
      ],
      Array [
        -122.41976308495022,
        37.80165833746338,
      ],
      Array [
        -122.41974046645252,
        37.801707302767504,
      ],
      Array [
        -122.41971810742356,
        37.80175722416558,
      ],
      Array [
        -122.41969595782884,
        37.8018079172897,
      ],
      Array [
        -122.41967396763383,
        37.80185919777198,
      ],
      Array [
        -122.41965208680396,
        37.801910881244474,
      ],
      Array [
        -122.41963026530469,
        37.80196278333927,
      ],
      Array [
        -122.41960845310155,
        37.80201471968846,
      ],
      Array [
        -122.41958660015999,
        37.80206650592411,
      ],
      Array [
        -122.41956465644546,
        37.802117957678334,
      ],
      Array [
        -122.41954257192343,
        37.80216889058318,
      ],
      Array [
        -122.41952029655936,
        37.80221912027076,
      ],
      Array [
        -122.41949778031874,
        37.80226846237314,
      ],
      Array [
        -122.41947497316703,
        37.802316732522414,
      ],
      Array [
        -122.41945182506966,
        37.80236374635065,
      ],
      Array [
        -122.41942828599221,
        37.802409319489946,
      ],
      Array [
        -122.41940430590003,
        37.80245326757239,
      ],
      Array [
        -122.4193798347586,
        37.80249540623004,
      ],
      Array [
        -122.41935482253346,
        37.80253555109499,
      ],
      Array [
        -122.41932921919002,
        37.80257351779934,
      ],
      Array [
        -122.41930297469379,
        37.802609121975166,
      ],
      Array [
        -122.41927603901021,
        37.80264217925455,
      ],
      Array [
        -122.41924836210475,
        37.80267250526956,
      ],
      Array [
        -122.41921989394287,
        37.8026999156523,
      ],
      Array [
        -122.41919058449007,
        37.80272422603484,
      ],
      Array [
        -122.4191603837118,
        37.802745252049284,
      ],
      Array [
        -122.41912924157351,
        37.80276280932769,
      ],
      Array [
        -122.4190971080407,
        37.80277671350214,
      ],
      Array [
        -122.4190639330788,
        37.80278678020474,
      ],
      Array [
        -122.41904497146605,
        37.802790657411244,
      ],
      Array [
        -122.41901000972386,
        37.80279401268719,
      ],
      Array [
        -122.41897382179135,
        37.80279260740665,
      ],
      Array [
        -122.41893646814607,
        37.802786683431094,
      ],
      Array [
        -122.41889800926566,
        37.802776482622036,
      ],
      Array [
        -122.41885850562758,
        37.80276224684095,
      ],
      Array [
        -122.41881801770954,
        37.80274421794934,
      ],
      Array [
        -122.41877660598905,
        37.80272263780869,
      ],
      Array [
        -122.4187343309437,
        37.8026977482805,
      ],
      Array [
        -122.41869125305111,
        37.80266979122625,
      ],
      Array [
        -122.4186474327888,
        37.80263900850743,
      ],
      Array [
        -122.4186029306344,
        37.80260564198556,
      ],
      Array [
        -122.41855780706545,
        37.80256993352211,
      ],
      Array [
        -122.41851212255956,
        37.80253212497857,
      ],
      Array [
        -122.41846593759433,
        37.80249245821643,
      ],
      Array [
        -122.41841931264727,
        37.802451175097204,
      ],
      Array [
        -122.41837230819606,
        37.80240851748236,
      ],
      Array [
        -122.4183249847182,
        37.8023647272334,
      ],
      Array [
        -122.41827740269127,
        37.80232004621181,
      ],
      Array [
        -122.41822962259292,
        37.802274716279086,
      ],
      Array [
        -122.41818170490068,
        37.802228979296714,
      ],
      Array [
        -122.41813371009214,
        37.80218307712621,
      ],
      Array [
        -122.41808569864489,
        37.80213725162904,
      ],
      Array [
        -122.41803773103652,
        37.802091744666704,
      ],
      Array [
        -122.41798986774458,
        37.802046798100704,
      ],
      Array [
        -122.41794216924666,
        37.80200265379251,
      ],
      Array [
        -122.41789469602033,
        37.80195955360362,
      ],
      Array [
        -122.4178475085432,
        37.801917739395556,
      ],
      Array [
        -122.41780066729288,
        37.80187745302976,
      ],
      Array [
        -122.41775423274687,
        37.801838936367766,
      ],
      Array [
        -122.41770826538279,
        37.80180243127104,
      ],
      Array [
        -122.41766282567824,
        37.80176817960109,
      ],
      Array [
        -122.41761797411077,
        37.801736423219396,
      ],
      Array [
        -122.41757377115798,
        37.80170740398746,
      ],
      Array [
        -122.41753027729746,
        37.80168136376677,
      ],
      Array [
        -122.41748755300677,
        37.80165854441882,
      ],
      Array [
        -122.41744565876347,
        37.80163918780509,
      ],
      Array [
        -122.41740465504519,
        37.80162353578708,
      ],
      Array [
        -122.4173646023295,
        37.80161183022629,
      ],
      Array [
        -122.41732556109396,
        37.80160431298421,
      ],
      Array [
        -122.41728759181616,
        37.801601225922326,
      ],
      Array [
        -122.4172507549737,
        37.80160281090213,
      ],
      Array [
        -122.41724252700804,
        37.801603850614384,
      ],
      Array [
        -122.41720693675963,
        37.80161067272053,
      ],
      Array [
        -122.417172100407,
        37.801620797280215,
      ],
      Array [
        -122.41713798820535,
        37.80163409399563,
      ],
      Array [
        -122.4171045704098,
        37.80165043256899,
      ],
      Array [
        -122.41707181727556,
        37.80166968270251,
      ],
      Array [
        -122.41703969905772,
        37.80169171409837,
      ],
      Array [
        -122.41700818601153,
        37.8017163964588,
      ],
      Array [
        -122.41697724839206,
        37.80174359948599,
      ],
      Array [
        -122.4169468564545,
        37.801773192882145,
      ],
      Array [
        -122.41691698045402,
        37.80180504634949,
      ],
      Array [
        -122.41688759064576,
        37.8018390295902,
      ],
      Array [
        -122.41685865728488,
        37.801875012306496,
      ],
      Array [
        -122.41683015062654,
        37.80191286420059,
      ],
      Array [
        -122.41680204092589,
        37.80195245497468,
      ],
      Array [
        -122.41677429843809,
        37.80199365433096,
      ],
      Array [
        -122.41674689341828,
        37.802036331971635,
      ],
      Array [
        -122.41671979612167,
        37.802080357598946,
      ],
      Array [
        -122.41669297680338,
        37.80212560091506,
      ],
      Array [
        -122.41666640571854,
        37.80217193162219,
      ],
      Array [
        -122.41664005312235,
        37.802219219422554,
      ],
      Array [
        -122.41661388926997,
        37.80226733401834,
      ],
      Array [
        -122.41658788441654,
        37.80231614511177,
      ],
      Array [
        -122.41656200881721,
        37.80236552240503,
      ],
      Array [
        -122.41653623272713,
        37.802415335600344,
      ],
      Array [
        -122.4165105264015,
        37.8024654543999,
      ],
      Array [
        -122.41648486009544,
        37.80251574850592,
      ],
      Array [
        -122.41645920406411,
        37.8025660876206,
      ],
      Array [
        -122.41643352856268,
        37.80261634144613,
      ],
      Array [
        -122.4164078038463,
        37.80266637968475,
      ],
      Array [
        -122.41638200017012,
        37.80271607203864,
      ],
      Array [
        -122.41635608778932,
        37.80276528821001,
      ],
      Array [
        -122.41633003695902,
        37.80281389790107,
      ],
      Array [
        -122.41630381793443,
        37.80286177081402,
      ],
      Array [
        -122.41627740097067,
        37.802908776651066,
      ],
      Array [
        -122.41625075632288,
        37.80295478511441,
      ],
      Array [
        -122.41622385424625,
        37.802999665906256,
      ],
      Array [
        -122.41619666499595,
        37.80304328872882,
      ],
      Array [
        -122.41616915882707,
        37.803085523284295,
      ],
      Array [
        -122.41614130599486,
        37.803126239274896,
      ],
      Array [
        -122.41612672805786,
        37.80314669573162,
      ],
    ],
    "type": "LineString",
  },
  "properties": Object {},
  "type": "Feature",
}
`;
